File: clients/fsharp/Model/ActionOrder.fs
#nowarn "0058"

namespace AiCup22
namespace AiCup22.Model

open AiCup22

/// Pick up loot
type ActionOrderPickup = {
    /// Loot id
    Loot: int;
} with

    /// Write Pickup to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 0
        writer.Write this.Loot
        ()

    /// Read Pickup from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Loot = reader.ReadInt32()
    }

/// Use shield potion
type ActionOrderUseShieldPotion = struct end with

    /// Write UseShieldPotion to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 1
        ()

    /// Read UseShieldPotion from reader
    static member readFrom(reader: System.IO.BinaryReader) = new ActionOrderUseShieldPotion()

/// Drop shield potions on the ground
type ActionOrderDropShieldPotions = {
    /// Amount of potions
    Amount: int;
} with

    /// Write DropShieldPotions to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 2
        writer.Write this.Amount
        ()

    /// Read DropShieldPotions from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Amount = reader.ReadInt32()
    }

/// Drop current weapon
type ActionOrderDropWeapon = struct end with

    /// Write DropWeapon to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 3
        ()

    /// Read DropWeapon from reader
    static member readFrom(reader: System.IO.BinaryReader) = new ActionOrderDropWeapon()

/// Drop ammo
type ActionOrderDropAmmo = {
    /// Weapon type index (starting with 0)
    WeaponTypeIndex: int;
    /// Amount of ammo
    Amount: int;
} with

    /// Write DropAmmo to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 4
        writer.Write this.WeaponTypeIndex
        writer.Write this.Amount
        ()

    /// Read DropAmmo from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        WeaponTypeIndex = reader.ReadInt32()
        Amount = reader.ReadInt32()
    }

/// Start/continue aiming
type ActionOrderAim = {
    /// Shoot (only possible in full aim)
    Shoot: bool;
} with

    /// Write Aim to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 5
        writer.Write this.Shoot
        ()

    /// Read Aim from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Shoot = reader.ReadBoolean()
    }

/// Order to perform an action for unit
type ActionOrder =
    /// Pick up loot
    | Pickup of ActionOrderPickup
    /// Use shield potion
    | UseShieldPotion of ActionOrderUseShieldPotion
    /// Drop shield potions on the ground
    | DropShieldPotions of ActionOrderDropShieldPotions
    /// Drop current weapon
    | DropWeapon of ActionOrderDropWeapon
    /// Drop ammo
    | DropAmmo of ActionOrderDropAmmo
    /// Start/continue aiming
    | Aim of ActionOrderAim
    with

    /// Write ActionOrder to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        match this with
            | Pickup value -> value.writeTo writer
            | UseShieldPotion value -> value.writeTo writer
            | DropShieldPotions value -> value.writeTo writer
            | DropWeapon value -> value.writeTo writer
            | DropAmmo value -> value.writeTo writer
            | Aim value -> value.writeTo writer

    /// Read ActionOrder from reader
    static member readFrom(reader: System.IO.BinaryReader) =
        match reader.ReadInt32() with
            | 0 -> Pickup (ActionOrderPickup.readFrom reader)
            | 1 -> UseShieldPotion (ActionOrderUseShieldPotion.readFrom reader)
            | 2 -> DropShieldPotions (ActionOrderDropShieldPotions.readFrom reader)
            | 3 -> DropWeapon (ActionOrderDropWeapon.readFrom reader)
            | 4 -> DropAmmo (ActionOrderDropAmmo.readFrom reader)
            | 5 -> Aim (ActionOrderAim.readFrom reader)
            | x -> failwith (sprintf "Unexpected tag %d" x)

File: clients/csharp/Model/ActionOrder.cs
namespace AiCup22.Model
{
    /// <summary>
    /// Order to perform an action for unit
    /// </summary>
    public abstract class ActionOrder
    {
        /// <summary> Write ActionOrder to writer </summary>
        public abstract void WriteTo(System.IO.BinaryWriter writer);

        /// <summary> Read ActionOrder from reader </summary>
        public static ActionOrder ReadFrom(System.IO.BinaryReader reader)
        {
            switch (reader.ReadInt32())
            {
                case Pickup.TAG:
                    return Pickup.ReadFrom(reader);
                case UseShieldPotion.TAG:
                    return UseShieldPotion.ReadFrom(reader);
                case DropShieldPotions.TAG:
                    return DropShieldPotions.ReadFrom(reader);
                case DropWeapon.TAG:
                    return DropWeapon.ReadFrom(reader);
                case DropAmmo.TAG:
                    return DropAmmo.ReadFrom(reader);
                case Aim.TAG:
                    return Aim.ReadFrom(reader);
                default:
                    throw new System.Exception("Unexpected tag value");
            }
        }

        /// <summary>
        /// Pick up loot
        /// </summary>
        public class Pickup : ActionOrder
        {
            public const int TAG = 0;
        
            /// <summary>
            /// Loot id
            /// </summary>
            public int Loot { get; set; }
        
            public Pickup() { }
        
            public Pickup(int loot)
            {
                this.Loot = loot;
            }
        
            /// <summary> Read Pickup from reader </summary>
            public static new Pickup ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new Pickup();
                result.Loot = reader.ReadInt32();
                return result;
            }
        
            /// <summary> Write Pickup to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                writer.Write(Loot);
            }
        
            /// <summary> Get string representation of Pickup </summary>
            public override string ToString() {
                string stringResult = "Pickup { ";
                stringResult += "Loot: ";
                stringResult += Loot.ToString();
                stringResult += " }";
                return stringResult;
            }
        }

        /// <summary>
        /// Use shield potion
        /// </summary>
        public class UseShieldPotion : ActionOrder
        {
            public const int TAG = 1;
        
        
            public UseShieldPotion() { }
        
            /// <summary> Read UseShieldPotion from reader </summary>
            public static new UseShieldPotion ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new UseShieldPotion();
                return result;
            }
        
            /// <summary> Write UseShieldPotion to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
            }
        
            /// <summary> Get string representation of UseShieldPotion </summary>
            public override string ToString() {
                string stringResult = "UseShieldPotion { ";
                stringResult += " }";
                return stringResult;
            }
        }

        /// <summary>
        /// Drop shield potions on the ground
        /// </summary>
        public class DropShieldPotions : ActionOrder
        {
            public const int TAG = 2;
        
            /// <summary>
            /// Amount of potions
            /// </summary>
            public int Amount { get; set; }
        
            public DropShieldPotions() { }
        
            public DropShieldPotions(int amount)
            {
                this.Amount = amount;
            }
        
            /// <summary> Read DropShieldPotions from reader </summary>
            public static new DropShieldPotions ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new DropShieldPotions();
                result.Amount = reader.ReadInt32();
                return result;
            }
        
            /// <summary> Write DropShieldPotions to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                writer.Write(Amount);
            }
        
            /// <summary> Get string representation of DropShieldPotions </summary>
            public override string ToString() {
                string stringResult = "DropShieldPotions { ";
                stringResult += "Amount: ";
                stringResult += Amount.ToString();
                stringResult += " }";
                return stringResult;
            }
        }

        /// <summary>
        /// Drop current weapon
        /// </summary>
        public class DropWeapon : ActionOrder
        {
            public const int TAG = 3;
        
        
            public DropWeapon() { }
        
            /// <summary> Read DropWeapon from reader </summary>
            public static new DropWeapon ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new DropWeapon();
                return result;
            }
        
            /// <summary> Write DropWeapon to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
            }
        
            /// <summary> Get string representation of DropWeapon </summary>
            public override string ToString() {
                string stringResult = "DropWeapon { ";
                stringResult += " }";
                return stringResult;
            }
        }

        /// <summary>
        /// Drop ammo
        /// </summary>
        public class DropAmmo : ActionOrder
        {
            public const int TAG = 4;
        
            /// <summary>
            /// Weapon type index (starting with 0)
            /// </summary>
            public int WeaponTypeIndex { get; set; }
            /// <summary>
            /// Amount of ammo
            /// </summary>
            public int Amount { get; set; }
        
            public DropAmmo() { }
        
            public DropAmmo(int weaponTypeIndex, int amount)
            {
                this.WeaponTypeIndex = weaponTypeIndex;
                this.Amount = amount;
            }
        
            /// <summary> Read DropAmmo from reader </summary>
            public static new DropAmmo ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new DropAmmo();
                result.WeaponTypeIndex = reader.ReadInt32();
                result.Amount = reader.ReadInt32();
                return result;
            }
        
            /// <summary> Write DropAmmo to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                writer.Write(WeaponTypeIndex);
                writer.Write(Amount);
            }
        
            /// <summary> Get string representation of DropAmmo </summary>
            public override string ToString() {
                string stringResult = "DropAmmo { ";
                stringResult += "WeaponTypeIndex: ";
                stringResult += WeaponTypeIndex.ToString();
                stringResult += ", ";
                stringResult += "Amount: ";
                stringResult += Amount.ToString();
                stringResult += " }";
                return stringResult;
            }
        }

        /// <summary>
        /// Start/continue aiming
        /// </summary>
        public class Aim : ActionOrder
        {
            public const int TAG = 5;
        
            /// <summary>
            /// Shoot (only possible in full aim)
            /// </summary>
            public bool Shoot { get; set; }
        
            public Aim() { }
        
            public Aim(bool shoot)
            {
                this.Shoot = shoot;
            }
        
            /// <summary> Read Aim from reader </summary>
            public static new Aim ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new Aim();
                result.Shoot = reader.ReadBoolean();
                return result;
            }
        
            /// <summary> Write Aim to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                writer.Write(Shoot);
            }
        
            /// <summary> Get string representation of Aim </summary>
            public override string ToString() {
                string stringResult = "Aim { ";
                stringResult += "Shoot: ";
                stringResult += Shoot.ToString();
                stringResult += " }";
                return stringResult;
            }
        }
    }
}